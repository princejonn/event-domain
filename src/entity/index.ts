export * from "./Aggregate";
export * from "./Saga";
export * from "./View";
