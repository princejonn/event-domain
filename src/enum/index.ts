export * from "./EventField";
export * from "./EventStoreField";
export * from "./Evt";
export * from "./MessageBusType";
export * from "./MessageType";
export * from "./MetaState";
export * from "./SagaStoreField";
export * from "./StoreType";
export * from "./ViewStoreField";
