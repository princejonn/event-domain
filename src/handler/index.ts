export * from "./AggregateCommandHandler";
export * from "./AggregateEventHandler";
export * from "./SagaEventHandler";
export * from "./ViewEventHandler";
