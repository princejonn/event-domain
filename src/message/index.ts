export * from "./Command";
export * from "./DomainEvent";
export * from "./Message";
export * from "./Timeout";
