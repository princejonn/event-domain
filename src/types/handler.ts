export interface HandlerIdentifier {
  name: string;
  context: string;
}

export interface HandlerConditions {
  created?: boolean;
  permanent?: boolean;
}

export interface HandlerIdentifierExpectingStructure {
  name?: string;
  context?: string;
}

export interface HandlerIdentifierMultipleContexts {
  name: string;
  context?: Array<string> | string;
}
