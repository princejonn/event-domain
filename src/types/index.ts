export * from "./aggregate";
export * from "./aggregate-command-handler";
export * from "./aggregate-domain";
export * from "./aggregate-event-handler";
export * from "./app";
export * from "./error";
export * from "./event-emitter";
export * from "./event-store";
export * from "./generic";
export * from "./guid";
export * from "./handler";
export * from "./message";
export * from "./message-bus";
export * from "./mongo";
export * from "./saga";
export * from "./saga-domain";
export * from "./saga-event-handler";
export * from "./saga-store";
export * from "./scan";
export * from "./view";
export * from "./view-domain";
export * from "./view-event-handler";
export * from "./view-store";
