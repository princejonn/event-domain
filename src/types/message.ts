import { AggregateIdentifier } from "./aggregate";
import { MessageType } from "../enum";

export interface IMessage {
  aggregate: AggregateIdentifier;
  causationId: string;
  correlationId: string;
  data: Record<string, any>;
  delay: number;
  id: string;
  mandatory: boolean;
  name: string;
  timestamp: Date;
  type: MessageType;
}

export interface MessageOptions {
  aggregate: AggregateIdentifier;
  causationId?: string;
  correlationId?: string;
  data?: Record<string, any>;
  delay?: number;
  id?: string;
  mandatory?: boolean;
  name: string;
  timestamp?: Date | string;
  type?: MessageType;
}
