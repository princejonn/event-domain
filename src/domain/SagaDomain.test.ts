import MockDate from "mockdate";
import { SagaDomain } from "./SagaDomain";
import { Saga } from "../entity";
import {
  aggregateIdentifier,
  commandWithCreated,
  commandWithData,
  eventWithData,
  eventWithDispatch,
  eventWithSaveOptions,
  logger,
  sagaHandlerWithData,
  sagaHandlerWithDispatch,
  sagaHandlerWithMultipleContext,
  sagaHandlerWithSaveOptions,
  sagaIdentifier,
} from "../test";
import { SagaEventHandler } from "../handler";
import { Command, Message } from "../message";
import {
  DomainError,
  HandlerNotRegisteredError,
  SagaAlreadyCreatedError,
  SagaNotCreatedError,
} from "../error";
import { MessageBusType, StoreType } from "../enum";
import { Subscription } from "../types";

MockDate.set("2022-01-01T08:00:00.000Z");

jest.mock("crypto", () => ({
  randomUUID: jest.fn(() => "e64b576a-022e-4dd4-8c97-32408012f8d1"),
}));

describe("SagaDomain", () => {
  describe("when registering an event handler", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      await domain.registerEventHandler(sagaHandlerWithData);
    });

    it("should become part of the domain", () => {
      // @ts-ignore
      expect(domain.eventHandlers).toMatchSnapshot();
    });

    it("should add subscriptions to topic", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].topic).toMatchSnapshot();
    });

    it("should add event subscription", async () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].sub).toMatchSnapshot();
    });

    it("should add handler callback", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].callback).toStrictEqual(
        expect.any(Function),
      );
    });
  });

  describe("when registering an event handler with multiple contexts", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      await domain.registerEventHandler(sagaHandlerWithMultipleContext);
    });

    it("should become part of the domain for each context", () => {
      // @ts-ignore
      expect(domain.eventHandlers).toMatchSnapshot();
    });

    it("should add subscriptions to topic for each context", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].topic).toMatchSnapshot();
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[1].topic).toMatchSnapshot();
    });

    it("should add event subscription for each context", async () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].sub).toMatchSnapshot();
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[1].sub).toMatchSnapshot();
    });

    it("should add handler callback for each context", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[0].callback).toStrictEqual(
        expect.any(Function),
      );
      // @ts-ignore
      expect(domain.messageBus.messageBus.subscriptions[1].callback).toStrictEqual(
        expect.any(Function),
      );
    });
  });

  describe("when registering an existing event handler", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      await domain.registerEventHandler(sagaHandlerWithData);
    });

    it("should throw", async () => {
      await expect(domain.registerEventHandler(sagaHandlerWithData)).rejects.toThrow();
    });
  });

  describe("when handling an event", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        getSagaId: () => "customEventId",
        handler: async (ctx) => {
          ctx.mergeState({ customEvent: true });
        },
      });
      await domain.registerEventHandler(customHandler);

      // @ts-ignore
      await domain.handleEvent(eventWithData, sagaIdentifier);
    });

    it("should use getSagaId on handler to generate id", () => {
      expect(
        // @ts-ignore
        domain.sagaStore.sagaStore.sagas["defaultContext.sagaName.customEventId"].id,
      ).toStrictEqual("customEventId");
    });
  });

  describe("when handling an event on a new saga", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      await domain.registerEventHandler(sagaHandlerWithDispatch);

      // @ts-ignore
      await domain.handleEvent(eventWithDispatch, sagaIdentifier);
    });

    it("should save the dispatched commands to saga store", () => {
      expect(
        // @ts-ignore
        domain.sagaStore.sagaStore.saveAttempts[0].saga.messagesToDispatch[0],
      ).toMatchSnapshot();
    });

    it("should publish the dispatched commands to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published[0]).toMatchSnapshot();
    });

    it("should clear the dispatched commands from saga store", () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.clearAttempts[0]).toStrictEqual(expect.any(Saga));
    });

    it("should resolve the promise with a saga that has a cleared list of commands", () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.sagas).toMatchSnapshot();
    });
  });

  describe("when handling an event with store options", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      await domain.registerEventHandler(sagaHandlerWithSaveOptions);

      // @ts-ignore
      await domain.handleEvent(eventWithSaveOptions, sagaIdentifier);
    });

    it("should pass the store options to the saga store", () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.saveAttempts[0].options).toStrictEqual({
        causationsCap: 100,
      });
    });
  });

  describe("when handling an event twice", () => {
    let domain: SagaDomain<any>;

    const handlerSpy = jest.fn();

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      const customHandler = new SagaEventHandler({
        eventName: eventWithDispatch.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        getSagaId: (event) => event.aggregate.id,
        handler: async (ctx) => {
          handlerSpy();
          ctx.dispatch(
            new Command({
              name: "commandDispatchFromHandler",
              aggregate: aggregateIdentifier,
            }),
          );
        },
      });

      await domain.registerEventHandler(customHandler);

      // @ts-ignore
      await domain.handleEvent(eventWithDispatch, sagaIdentifier);
      // @ts-ignore
      await domain.handleEvent(eventWithDispatch, sagaIdentifier);
    });

    it("should call the event handler once", () => {
      expect(handlerSpy).toHaveBeenCalledTimes(1);
    });

    it("should save the dispatched commands to saga store once", () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.saveAttempts.length).toBe(1);
    });

    it("should publish the dispatched commands to message bus once", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published.length).toBe(1);
    });

    it("should clear the dispatched commands to saga store once", () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.clearAttempts.length).toBe(1);
    });
  });

  describe("when handling an event and failing to publish commands", () => {
    let test: any = {};
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({
        messageBus: {
          type: MessageBusType.CUSTOM,
          custom: {
            subscribe: async (subscription: Subscription) => {
              test.subscribe = { subscription };
            },
            publish: async (messages: Array<Message>) => {
              test.publish = { messages };
              throw new Error("custom");
            },
          },
        },
        sagaStore: {},
        logger,
      });

      await domain.registerEventHandler(sagaHandlerWithDispatch);
    });

    it("should reject command", async () => {
      // @ts-ignore
      await expect(domain.handleEvent(eventWithDispatch, sagaIdentifier)).rejects.toThrow(
        new Error("custom"),
      );
    });

    it("should save dispatched commands to saga store", async () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.saveAttempts).toMatchSnapshot();
    });

    it("should publish the dispatched commands to message bus", async () => {
      expect(test.publish).toMatchSnapshot();
    });
  });

  describe("when handling an event for a destroyed saga", () => {
    let test: any = {};
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({
        sagaStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async (sagaIdentifier) => {
              test.load = { sagaIdentifier };
              return new Saga<any>(
                {
                  id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
                  name: "sagaName",
                  context: "defaultContext",
                  destroyed: true,
                },
                logger,
              );
            },
            save: async (saga, causation, options?) => {
              test.save = { saga, causation, options };
              return saga;
            },
            clearMessagesToDispatch: async (saga) => {
              test.clearMessagesToDispatch = { saga };
              return new Saga({ ...saga, messagesToDispatch: [] }, logger);
            },
          },
        },
        messageBus: {},
        logger,
      });

      await domain.registerEventHandler(sagaHandlerWithData);

      // @ts-ignore
      await domain.handleEvent(eventWithData, sagaIdentifier);
    });

    it("should not save the saga", () => {
      expect(test.save).toBeUndefined();
    });
  });

  describe("when handling an event with created [true] condition", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        conditions: {
          created: true,
        },
        getSagaId: () => "customEventId",
        handler: async (ctx) => {
          ctx.mergeState({ customEvent: true });
        },
      });
      await domain.registerEventHandler(customHandler);
    });

    it("should reject the promise with a transient SagaNotCreatedError", async () => {
      // @ts-ignore
      await expect(domain.handleEvent(eventWithData, sagaIdentifier)).rejects.toThrow(
        SagaNotCreatedError,
      );
    });
  });

  describe("when handling an event with created [true] and permanent [true] condition", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        conditions: {
          created: true,
          permanent: true,
        },
        getSagaId: () => "customEventId",
        handler: async (ctx) => {
          ctx.mergeState({ customEvent: true });
        },
      });

      await domain.registerEventHandler(customHandler);
      await domain.registerEventHandler(sagaHandlerWithDispatch);

      // @ts-ignore
      await domain.handleEvent(eventWithDispatch, sagaIdentifier);
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(eventWithData, sagaIdentifier),
      ).resolves.toBeUndefined();
    });

    it("should not save the saga", () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.saveAttempts.length).toBe(1);
    });

    it("should publish SagaNotCreatedError", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published).toStrictEqual(
        expect.arrayContaining([
          expect.objectContaining({
            name: "SagaNotCreatedError",
          }),
        ]),
      );
    });
  });

  describe("when handling an event with created [false] condition", () => {
    let test: any = {};
    let domain: SagaDomain<any>;

    const clearSpy = jest.fn();

    beforeAll(async () => {
      domain = new SagaDomain({
        sagaStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async (sagaIdentifier) => {
              test.load = { sagaIdentifier };
              return new Saga<any>(
                {
                  id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
                  name: "sagaName",
                  context: "defaultContext",
                  messagesToDispatch: [commandWithData],
                  revision: 1,
                },
                logger,
              );
            },
            save: async (saga, causation, options?) => {
              test.save = { saga, causation, options };
              return saga;
            },
            clearMessagesToDispatch: async (saga) => {
              clearSpy(saga);
              test.clearMessagesToDispatch = { saga };
              return new Saga({ ...saga, messagesToDispatch: [] }, logger);
            },
          },
        },
        messageBus: {},
        logger,
      });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        conditions: {
          created: false,
        },
        getSagaId: () => "customEventId",
        handler: async (ctx) => {
          ctx.mergeState({ customEvent: true });
        },
      });

      await domain.registerEventHandler(customHandler);
    });

    it("should resolve", async () => {
      await expect(
        // @ts-ignore
        domain.handleEvent(eventWithData, sagaIdentifier),
      ).resolves.toBeUndefined();
    });

    it("should not save the saga", () => {
      expect(test.save).toBeUndefined();
    });

    it("should try to clear commands from saga", () => {
      expect(clearSpy).toHaveBeenCalledWith(
        expect.objectContaining({
          messagesToDispatch: expect.arrayContaining([commandWithData]),
        }),
      );
    });

    it("publish undispatched commands to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published).toStrictEqual(
        expect.arrayContaining([commandWithData]),
      );
    });

    it("publish SagaAlreadyCreatedError to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published).toStrictEqual(
        expect.arrayContaining([
          expect.objectContaining({
            name: "SagaAlreadyCreatedError",
          }),
        ]),
      );
    });
  });

  describe("when handling an event with created [false] and permanent [false] condition", () => {
    let test: any = {};
    let domain: SagaDomain<any>;

    const clearSpy = jest.fn();

    beforeAll(async () => {
      domain = new SagaDomain({
        sagaStore: {
          type: StoreType.CUSTOM,
          custom: {
            load: async (sagaIdentifier) => {
              test.load = { sagaIdentifier };
              return new Saga<any>(
                {
                  id: "2a553a74-0f84-4a7d-b91e-2902666db98c",
                  name: "sagaName",
                  context: "defaultContext",
                  messagesToDispatch: [commandWithData],
                  revision: 1,
                },
                logger,
              );
            },
            save: async (saga, causation, options?) => {
              test.save = { saga, causation, options };
              return saga;
            },
            clearMessagesToDispatch: async (saga) => {
              clearSpy(saga);
              test.clearMessagesToDispatch = { saga };
              return new Saga({ ...saga, messagesToDispatch: [] }, logger);
            },
          },
        },
        messageBus: {},
        logger,
      });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        conditions: {
          created: false,
          permanent: false,
        },
        getSagaId: () => "customEventId",
        handler: async (ctx) => {
          ctx.mergeState({ customEvent: true });
        },
      });

      await domain.registerEventHandler(customHandler);
    });

    it("should reject with transient SagaAlreadyCreatedError", async () => {
      // @ts-ignore
      await expect(domain.handleEvent(eventWithData, sagaIdentifier)).rejects.toThrow(
        SagaAlreadyCreatedError,
      );
    });
  });

  describe("when no event handler has been registered", () => {
    let domain: SagaDomain<any>;

    beforeAll(async () => {
      domain = new SagaDomain({ sagaStore: {}, messageBus: {}, logger });

      await domain.registerEventHandler(sagaHandlerWithDispatch);
      await domain.registerEventHandler(sagaHandlerWithSaveOptions);
    });

    it("should throw HandlerNotRegisteredError", async () => {
      // @ts-ignore
      await expect(domain.handleEvent(eventWithData, sagaIdentifier)).rejects.toThrow(
        HandlerNotRegisteredError,
      );
    });
  });

  describe("when event handler throws an error", () => {
    let domain: SagaDomain<any>;

    const saveSpy = jest.fn();
    const clearSpy = jest.fn();

    beforeAll(async () => {
      domain = new SagaDomain({
        sagaStore: {
          type: StoreType.CUSTOM,
          custom: {
            // @ts-ignore
            save: () => {
              saveSpy();
            },
            // @ts-ignore
            load: () => {
              return new Saga(
                {
                  ...sagaIdentifier,
                  messagesToDispatch: [commandWithCreated],
                },
                logger,
              );
            },
            // @ts-ignore
            clearMessagesToDispatch: () => {
              clearSpy();
            },
          },
        },
        messageBus: {},
        logger,
      });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        getSagaId: (event) => event.aggregate.id,
        handler: async () => {
          throw new Error("HandlerError");
        },
      });

      await domain.registerEventHandler(customHandler);
    });

    it("should throw the error", async () => {
      // @ts-ignore
      await expect(domain.handleEvent(eventWithData, sagaIdentifier)).rejects.toThrow(
        new Error("HandlerError"),
      );
    });

    it("should not save the saga", () => {
      expect(saveSpy).not.toHaveBeenCalled();
    });

    it("should not publish any undispatched commands on the saga", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published.length).toBe(0);
    });

    it("should not clear any undispatched commands on the saga", () => {
      expect(clearSpy).not.toHaveBeenCalled();
    });
  });

  describe("when event handler throws a permanent domain error", () => {
    let domain: SagaDomain<any>;

    class TestError extends DomainError {
      constructor() {
        super("error", { permanent: true });
      }
    }

    beforeAll(async () => {
      domain = new SagaDomain({
        sagaStore: {},
        messageBus: {},
        logger,
      });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        getSagaId: (event) => event.aggregate.id,
        handler: async () => {
          throw new TestError();
        },
      });

      await domain.registerEventHandler(customHandler);
    });

    it("should resolve", async () => {
      // @ts-ignore
      await expect(domain.handleEvent(eventWithData, sagaIdentifier)).resolves.toBe(
        undefined,
      );
    });

    it("should not save the saga", () => {
      // @ts-ignore
      expect(domain.sagaStore.sagaStore.saveAttempts.length).toBe(0);
    });

    it("should publish a non-mandatory error event to message bus", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published[0]).toMatchSnapshot();
    });
  });

  describe("when event handler throws a non-permanent domain error", () => {
    let domain: SagaDomain<any>;

    class TestError extends DomainError {
      constructor() {
        super("error", { permanent: false });
      }
    }

    const saveSpy = jest.fn();
    const clearSpy = jest.fn();

    beforeAll(async () => {
      domain = new SagaDomain({
        sagaStore: {
          type: StoreType.CUSTOM,
          custom: {
            // @ts-ignore
            save: () => {
              saveSpy();
            },
            // @ts-ignore
            load: () => {
              return new Saga(
                {
                  ...sagaIdentifier,
                  messagesToDispatch: [commandWithCreated],
                },
                logger,
              );
            },
            // @ts-ignore
            clearMessagesToDispatch: () => {
              clearSpy();
            },
          },
        },
        messageBus: {},
        logger,
      });

      const customHandler = new SagaEventHandler({
        eventName: eventWithData.name,
        aggregate: aggregateIdentifier,
        saga: sagaIdentifier,
        getSagaId: (event) => event.aggregate.id,
        handler: async () => {
          throw new TestError();
        },
      });

      await domain.registerEventHandler(customHandler);
    });

    it("should throw the error", async () => {
      // @ts-ignore
      await expect(domain.handleEvent(eventWithData, sagaIdentifier)).rejects.toThrow(
        TestError,
      );
    });

    it("should not save the saga", () => {
      expect(saveSpy).not.toHaveBeenCalled();
    });

    it("should not publish any undispatched commands on the saga", () => {
      // @ts-ignore
      expect(domain.messageBus.messageBus.published.length).toBe(0);
    });

    it("should not clear any undispatched commands on the saga", () => {
      expect(clearSpy).not.toHaveBeenCalled();
    });
  });
});
