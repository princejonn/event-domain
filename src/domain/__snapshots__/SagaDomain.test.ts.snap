// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SagaDomain when event handler throws a permanent domain error should publish a non-mandatory error event to message bus 1`] = `
DomainEvent {
  "aggregate": Object {
    "context": "defaultContext",
    "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
    "name": "sagaName",
  },
  "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "data": Object {
    "error": [TestError: error],
    "message": DomainEvent {
      "aggregate": Object {
        "context": "defaultContext",
        "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
        "name": "aggregateName",
      },
      "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "data": Object {
        "name": "hello",
      },
      "delay": 0,
      "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "mandatory": false,
      "name": "eventWithData",
      "timestamp": 2020-01-01T12:00:00.000Z,
      "type": "event",
    },
  },
  "delay": 0,
  "id": "e64b576a-022e-4dd4-8c97-32408012f8d1.TestError",
  "mandatory": false,
  "name": "TestError",
  "timestamp": 2022-01-01T08:00:00.000Z,
  "type": "event",
}
`;

exports[`SagaDomain when handling an event and failing to publish commands should publish the dispatched commands to message bus 1`] = `
Object {
  "messages": Array [
    Command {
      "aggregate": Object {
        "context": "defaultContext",
        "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
        "name": "aggregateName",
      },
      "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "data": Object {},
      "delay": 0,
      "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "mandatory": true,
      "name": "commandDispatchFromHandler",
      "timestamp": 2022-01-01T08:00:00.000Z,
      "type": "command",
    },
  ],
}
`;

exports[`SagaDomain when handling an event and failing to publish commands should save dispatched commands to saga store 1`] = `
Array [
  Object {
    "causation": DomainEvent {
      "aggregate": Object {
        "context": "defaultContext",
        "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
        "name": "aggregateName",
      },
      "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "data": Object {},
      "delay": 0,
      "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
      "mandatory": false,
      "name": "eventWithDispatch",
      "timestamp": 2020-01-01T12:00:00.000Z,
      "type": "event",
    },
    "options": Object {},
    "saga": Object {
      "causationList": Array [
        "e64b576a-022e-4dd4-8c97-32408012f8d1",
      ],
      "context": "defaultContext",
      "destroyed": false,
      "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
      "messagesToDispatch": Array [
        Command {
          "aggregate": Object {
            "context": "defaultContext",
            "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
            "name": "aggregateName",
          },
          "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
          "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
          "data": Object {},
          "delay": 0,
          "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
          "mandatory": true,
          "name": "commandDispatchFromHandler",
          "timestamp": 2022-01-01T08:00:00.000Z,
          "type": "command",
        },
      ],
      "name": "sagaName",
      "revision": 0,
      "state": Object {},
    },
  },
]
`;

exports[`SagaDomain when handling an event on a new saga should publish the dispatched commands to message bus 1`] = `
Command {
  "aggregate": Object {
    "context": "defaultContext",
    "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
    "name": "aggregateName",
  },
  "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "data": Object {},
  "delay": 0,
  "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "mandatory": true,
  "name": "commandDispatchFromHandler",
  "timestamp": 2022-01-01T08:00:00.000Z,
  "type": "command",
}
`;

exports[`SagaDomain when handling an event on a new saga should resolve the promise with a saga that has a cleared list of commands 1`] = `
Object {
  "defaultContext.sagaName.2a553a74-0f84-4a7d-b91e-2902666db98c": Object {
    "causationList": Array [
      "e64b576a-022e-4dd4-8c97-32408012f8d1",
    ],
    "context": "defaultContext",
    "destroyed": false,
    "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
    "messagesToDispatch": Array [],
    "name": "sagaName",
    "revision": 2,
    "state": Object {},
  },
}
`;

exports[`SagaDomain when handling an event on a new saga should save the dispatched commands to saga store 1`] = `
Command {
  "aggregate": Object {
    "context": "defaultContext",
    "id": "2a553a74-0f84-4a7d-b91e-2902666db98c",
    "name": "aggregateName",
  },
  "causationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "correlationId": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "data": Object {},
  "delay": 0,
  "id": "e64b576a-022e-4dd4-8c97-32408012f8d1",
  "mandatory": true,
  "name": "commandDispatchFromHandler",
  "timestamp": 2022-01-01T08:00:00.000Z,
  "type": "command",
}
`;

exports[`SagaDomain when registering an event handler should add event subscription 1`] = `
Object {
  "aggregate": Object {
    "context": "defaultContext",
    "name": "sagaName",
  },
  "handler": [Function],
  "name": "defaultContext.aggregateName.eventWithData",
}
`;

exports[`SagaDomain when registering an event handler should add subscriptions to topic 1`] = `
Object {
  "aggregate": Object {
    "context": "defaultContext",
    "name": "aggregateName",
  },
  "handler": [Function],
  "name": "eventWithData",
}
`;

exports[`SagaDomain when registering an event handler should become part of the domain 1`] = `
Array [
  SagaEventHandler {
    "aggregate": Object {
      "context": "defaultContext",
      "name": "aggregateName",
    },
    "conditions": Object {},
    "eventName": "eventWithData",
    "getSagaId": [Function],
    "handler": [Function],
    "saga": Object {
      "context": "defaultContext",
      "name": "sagaName",
    },
    "saveOptions": Object {},
  },
]
`;

exports[`SagaDomain when registering an event handler with multiple contexts should add event subscription for each context 1`] = `
Object {
  "aggregate": Object {
    "context": "defaultContext",
    "name": "sagaName",
  },
  "handler": [Function],
  "name": "defaultContext.aggregateName.eventWithTimeout",
}
`;

exports[`SagaDomain when registering an event handler with multiple contexts should add event subscription for each context 2`] = `
Object {
  "aggregate": Object {
    "context": "defaultContext",
    "name": "sagaName",
  },
  "handler": [Function],
  "name": "anotherContext.aggregateName.eventWithTimeout",
}
`;

exports[`SagaDomain when registering an event handler with multiple contexts should add subscriptions to topic for each context 1`] = `
Object {
  "aggregate": Object {
    "context": "defaultContext",
    "name": "aggregateName",
  },
  "handler": [Function],
  "name": "eventWithTimeout",
}
`;

exports[`SagaDomain when registering an event handler with multiple contexts should add subscriptions to topic for each context 2`] = `
Object {
  "aggregate": Object {
    "context": "anotherContext",
    "name": "aggregateName",
  },
  "handler": [Function],
  "name": "eventWithTimeout",
}
`;

exports[`SagaDomain when registering an event handler with multiple contexts should become part of the domain for each context 1`] = `
Array [
  SagaEventHandler {
    "aggregate": Object {
      "context": "defaultContext",
      "name": "aggregateName",
    },
    "conditions": Object {},
    "eventName": "eventWithTimeout",
    "getSagaId": [Function],
    "handler": [Function],
    "saga": Object {
      "context": "defaultContext",
      "name": "sagaName",
    },
    "saveOptions": Object {},
  },
  SagaEventHandler {
    "aggregate": Object {
      "context": "anotherContext",
      "name": "aggregateName",
    },
    "conditions": Object {},
    "eventName": "eventWithTimeout",
    "getSagaId": [Function],
    "handler": [Function],
    "saga": Object {
      "context": "defaultContext",
      "name": "sagaName",
    },
    "saveOptions": Object {},
  },
]
`;
